[[ Bug 14422 ]] Allow specification of min_os_version fields when buildi...

runrevmark opened the pull request:
...ng mach-o standalones.

The plist in an app bundle must have a MinOsVersion field which matches the mach-o load command inside the executable.

This patch adds a 'min_os_version' key to the deploy array used when using the deploy command. The key's value can either be a string or an array.

In the former case, the string is taken to be a version to apply to all architectures in the (fat) binary.

In the latter case, the array is expected to be a mapping from architecture string (key) to version (value). The architecture string can be one of:
  i386, x86_64, armv6, armv7, armv7s, arm64, ppc, ppc64
If the architecture string is empty it is taken to be 'unknown'.

The version (value) must be in the X.Y.Z where X, Y, Z are integers.

For example:
   put "6.1.0" into tDeploy["min_os_version"]["i386"]
   put "8.1.0" into tDeploy["min_os_version"]["armv7"]
   put "7.1.0" into tDeploy["min_os_version"]["arm64"]

When building a standalone, the deploy command will rewrite the min os version load command in each slice based on the architecture of the slice. If a given architecture is not present in the min_os_version array, then it will take the value specified for the 'unknown' (empty key). If neither a given architecture is present nor an 'unknown' entry then the deploy command will not touch the min_os_version load command.

NOTE: This pull request should be augmented with the changes to the standalone builder to ensure that the min-os-version that is written in the plist matches that which is written into the binary.

runrevmark committed:
[[ Bug 14422 ]] Allow specification of min_os_version fields when building mach-o standalones.

The plist in an app bundle must have a MinOsVersion field which matches the mach-o load command inside the executable.

This patch adds a 'min_os_version' key to the deploy array used when using the deploy command. The key's value can either be a string or an array.

In the former case, the string is taken to be a version to apply to all architectures in the (fat) binary.

In the latter case, the array is expected to be a mapping from architecture string (key) to version (value). The architecture string can be one of:
  i386, x86_64, armv6, armv7, armv7s, arm64, ppc, ppc64
If the architecture string is empty it is taken to be 'unknown'.

The version (value) must be in the X.Y.Z where X, Y, Z are integers.

For example:
   put "6.1.0" into tDeploy["min_os_version"]["i386"]
   put "8.1.0" into tDeploy["min_os_version"]["armv7"]
   put "7.1.0" into tDeploy["min_os_version"]["arm64"]

When building a standalone, the deploy command will rewrite the min os version load command in each slice based on the architecture of the slice. If a given architecture is not present in the min_os_version array, then it will take the value specified for the 'unknown' (empty key). If neither a given architecture is present nor an 'unknown' entry then the deploy command will not touch the min_os_version load command.

peter-b commented on engine/src/deploy_macosx.cpp:
```diff
@@ -1088,6 +1135,55 @@ static void relocate_function_starts_command(linkedit_data_command *x, int32_t p
 
+static bool MCDeployToMacOSXFetchMinOSVersion(const MCDeployParameters& p_params, mach_header& p_header, uint32_t& r_version)
+{
+    // First work out what DeployArchitecture to look for.
```
Would this be more readable as a `switch` block? (Just wondering).

runrevmark replied:
It needs to be a cascaded if since some cpu-types depend on two fields to determine - however, I did notice that I wasn't setting 'Unknown' in the case of no match - so have fixed that.

peter-b commented on engine/src/deploy_macosx.cpp:
```diff
@@ -1088,6 +1135,55 @@ static void relocate_function_starts_command(linkedit_data_command *x, int32_t p
+    int t_unknown_index, t_found_index;
+    t_unknown_index = -1;
+    t_found_index = -1;
+    for(uindex_t i = 0; i < p_params . min_os_version_count; i++)
```
`min_os_version_count` has type `uint32_t`. But I think it's okay to use `uindex_t i` because we're planning to ensure `sizeof(uindex_t) > sizeof(uint32_t)`.

runrevmark replied:
I've changed the deploy parameters struct to use uindex_t. It uses MCMemoryResizeArray on the ptr, so its best to be consistent.

peter-b commented on engine/src/deploy_macosx.cpp:
```diff
@@ -1088,6 +1135,55 @@ static void relocate_function_starts_command(linkedit_data_command *x, int32_t p
+    t_found_index = -1;
+    for(uindex_t i = 0; i < p_params . min_os_version_count; i++)
+        if (p_params . min_os_versions[i] . architecture == t_arch &&
+            t_found_index == -1)
```
All these comparisons should probably be "is an invalid index" (i.e. `x < 0`) rather than `x == -1`.

runrevmark replied:
Agreed.

peter-b commented on engine/src/deploy.cpp:
```diff
@@ -531,6 +582,51 @@ Exec_stat MCIdeDeploy::exec(MCExecPoint& ep)
+    if (t_stat == ES_NORMAL)
+    {
+        t_stat = t_array -> fetch_element(ep2, "min_os_version");
+        if (t_stat == ES_NORMAL)
```
This doesn't seem to need to be nested.

runrevmark replied:
True - but it was more to make it clear what code is manipulating the 'min_os_version' field - its all contained in the outer { ... }.

peter-b commented on engine/src/deploy.cpp:
```diff
@@ -89,6 +89,57 @@ extern Boolean InitSSLCrypt(void);
+{
+    int t_major, t_minor, t_inc;
+    t_major = t_minor = t_inc = 0;
+    sscanf(p_vers_string, "%d.%d.%d", &t_major, &t_minor, &t_inc);
```
I'm assuming that you don't check the return value of `sscanf()` because if it returns < 3 then you just assume that some components of the version string were missing.  They get initialised anyway, after all! :wink:

runrevmark replied:
I've added a comment :)

runrevmark committed:
[[ Bug 14422 ]] Cleanup a few stylistic things.

peter-b commented:
Reviewed-By: Peter TB Brett peter@livecode.com

livecodesebastien committed:
[[ Bug 14422 ]] Fix few issues on the engine side, and add the IDE part of the minimal OS version

livecodesebastien committed:
[[ Bug 14422 ]] Change the minimal iOS system version to be architecture-agnostic.

livecodesebastien committed:
[[ Bug 14422 ]] Restore MinimalOSVersion in the Plist (LSMinimalVersion does not work)
                64-bit minimal OS version is 5.1.1, not 5.1.0

livecodesebastien committed:
[[ Bug 14422 ]] Update the ide submodule
** This change is only valid for release-6.7.2 as it points to livecode-ide/release-6.7.2 **

livecodesebastien committed:
[[ Bug 14422 ]] Re-update the ide submodule

livecodesebastien committed:
Revert "[[ Bug 14422 ]] Re-update the ide submodule"

This reverts commit ae121cbd5dbf23d99e940cf7f462d64fd41f8492.

livecodesebastien committed:
Revert "[[ Bug 14422 ]] Update the ide submodule"

This reverts commit 7fd7386155ca8ba12d1ba5671a60fdfc272277a0.

livecodesebastien committed:
[[ Bug 14422 ]] Added bug fix note

runrevmark commented:
Superceded.
